### PR TITLE
[Refactor] Extract ExecutionDAG and Deployer from Coordinator

### PR DESCRIPTION
**Pull request by ZiheLiu**
Relates to #27846.

The execution DAG represents the distributed execution of a job, mainly including parallel instances of fragments and the communication between them.
The execution DAG consists of the following components:
- The `ExecutionFragment` represents a collection of multiple parallel instances of a PlanFragment.
   Additionally, it includes information about the source and sink, including assigned scan ranges and destinations to the downstream.
- The `FragmentInstance` represents a parallel instance of a PlanFragment.
- The `FragmentInstanceExecState` represents the execution state of a FragmentInstance for a single attempt.

BTW, move the deployment logic to a utility class `Deployer`.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 646 / 759 (85.11%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstance.java|48|67|71.64%|[66, 67, 97, 104, 164, 165, 168, 169, 170, 173, 174, 175, 177, 178, 183, 191, 192, 193, 195]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionDAG.java|174|216|80.56%|[87, 91, 143, 160, 205, 226, 228, 229, 283, 284, 285, 351, 352, 353, 355, 356, 388, 404, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 478, 479, 480, 481, 482, 483, 487, 488, 489, 490, 491]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstanceExecState.java|5|6|83.33%|[390]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionFragment.java|146|170|85.88%|[94, 98, 102, 116, 139, 146, 262, 270, 274, 310, 311, 313, 314, 318, 319, 320, 324, 325, 326, 327, 341, 352, 353, 359]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|51|58|87.93%|[231, 235, 381, 382, 383, 487, 660]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|54|61|88.52%|[359, 628, 922, 923, 1039, 1125, 1359]|
|:large_blue_circle:|com/starrocks/qe/scheduler/TFragmentInstanceFactory.java|83|93|89.25%|[72, 121, 147, 148, 149, 150, 151, 173, 174, 175]|
|:large_blue_circle:|com/starrocks/qe/scheduler/Deployer.java|71|74|95.95%|[182, 183, 184]|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 675 / 800 (84.38%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstance.java|48|67|71.64%|[66, 67, 97, 104, 164, 165, 168, 169, 170, 173, 174, 175, 177, 178, 183, 191, 192, 193, 195]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionDAG.java|174|218|79.82%|[107, 108, 109, 174, 175, 176, 178, 179, 211, 227, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 274, 278, 314, 315, 316, 317, 318, 319, 323, 324, 325, 326, 327, 328, 329, 364, 381, 426, 447, 449, 450]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|73|89|82.02%|[231, 235, 236, 376, 377, 380, 381, 382, 383, 486, 487, 656, 686, 687, 706, 713]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstanceExecState.java|5|6|83.33%|[390]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionFragment.java|146|170|85.88%|[95, 99, 103, 119, 142, 149, 263, 264, 266, 267, 271, 272, 273, 277, 278, 279, 280, 294, 305, 306, 312, 342, 350, 354]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|54|61|88.52%|[359, 628, 922, 923, 1039, 1125, 1359]|
|:large_blue_circle:|com/starrocks/qe/scheduler/TFragmentInstanceFactory.java|83|93|89.25%|[72, 121, 147, 148, 149, 150, 151, 173, 174, 175]|
|:large_blue_circle:|com/starrocks/qe/scheduler/Deployer.java|78|82|95.12%|[168, 169, 170, 200]|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 676 / 801 (84.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstance.java|48|67|71.64%|[66, 67, 97, 104, 164, 165, 168, 169, 170, 173, 174, 175, 177, 178, 183, 191, 192, 193, 195]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionDAG.java|174|218|79.82%|[107, 108, 109, 174, 175, 176, 178, 179, 211, 227, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 274, 278, 314, 315, 316, 317, 318, 319, 323, 324, 325, 326, 327, 328, 329, 364, 381, 426, 447, 449, 450]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|73|89|82.02%|[231, 235, 236, 376, 377, 380, 381, 382, 383, 486, 487, 656, 686, 687, 706, 713]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstanceExecState.java|5|6|83.33%|[390]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionFragment.java|146|170|85.88%|[95, 99, 103, 119, 142, 149, 263, 264, 266, 267, 271, 272, 273, 277, 278, 279, 280, 294, 305, 306, 312, 342, 350, 354]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|55|62|88.71%|[359, 630, 924, 925, 1041, 1127, 1361]|
|:large_blue_circle:|com/starrocks/qe/scheduler/TFragmentInstanceFactory.java|83|93|89.25%|[72, 121, 147, 148, 149, 150, 151, 173, 174, 175]|
|:large_blue_circle:|com/starrocks/qe/scheduler/Deployer.java|78|82|95.12%|[168, 169, 170, 200]|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/Deployer.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        // This is a load process, and it is the first fragment.
+        // we should add all BackendExecState of this fragment to needCheckBackendExecStates,
+        // so that we can check these backends' state when joining this Coordinator
+        boolean needCheckExecutionState = jobSpec.isLoadType() && fragment.getFragmentIndex() == 0;
+        boolean isEnablePipeline = jobSpec.isEnablePipeline();
+        // if pipeline is enable and current fragment contain olap table sink, in fe we will
+        // calculate the number of all tablet sinks in advance and assign them to each fragment instance
+        boolean enablePipelineTableSinkDop = isEnablePipeline && fragment.getPlanFragment().hasTableSink();
+
+        List<List<FragmentInstance>> twoStageInstancesToDeploy = ImmutableList.of(new ArrayList<>(), new ArrayList<>());
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/qe/scheduler/Deployer.java:113]: what does two stage instance mean?

Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/qe/scheduler/Deployer.java:113]: This is explained in the comments in `deployFragments`:
```java
    public void deployFragments(List<ExecutionFragment> concurrentFragments) throws RpcException, UserException {
        // Divide requests of fragments in the current group to two stages.
        // - stage 1, the first request to a host, which need send descTable.
        // - stage 2, the non-first requests to a host, which needn't send descTable.
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 674 / 799 (84.36%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstance.java|48|67|71.64%|[79, 80, 110, 117, 177, 178, 181, 182, 183, 186, 187, 188, 190, 191, 196, 204, 205, 206, 208]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionDAG.java|174|218|79.82%|[130, 131, 132, 199, 200, 201, 203, 204, 236, 252, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 308, 312, 348, 349, 350, 351, 352, 353, 357, 358, 359, 360, 361, 362, 363, 398, 415, 460, 481, 483, 484]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|72|88|81.82%|[226, 230, 231, 371, 372, 375, 376, 377, 378, 481, 482, 651, 681, 682, 701, 708]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstanceExecState.java|5|6|83.33%|[390]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionFragment.java|142|166|85.54%|[103, 107, 111, 127, 150, 157, 262, 263, 265, 266, 270, 271, 272, 276, 277, 278, 279, 293, 304, 305, 311, 341, 349, 353]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|54|61|88.52%|[359, 635, 929, 930, 1046, 1132, 1366]|
|:large_blue_circle:|com/starrocks/qe/scheduler/TFragmentInstanceFactory.java|83|93|89.25%|[72, 121, 147, 148, 149, 150, 151, 173, 174, 175]|
|:large_blue_circle:|com/starrocks/qe/scheduler/Deployer.java|78|82|95.12%|[171, 172, 173, 203]|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|9|9|100.00%|[]|

**Review by liuyehcf — DISMISSED** [review]

**Comment by liuyehcf** [comment]
LGTM

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 674 / 799 (84.36%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstance.java|48|67|71.64%|[79, 80, 110, 117, 177, 178, 181, 182, 183, 186, 187, 188, 190, 191, 196, 204, 205, 206, 208]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionDAG.java|174|218|79.82%|[130, 131, 132, 199, 200, 201, 203, 204, 236, 252, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 308, 312, 348, 349, 350, 351, 352, 353, 357, 358, 359, 360, 361, 362, 363, 398, 415, 460, 481, 483, 484]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|72|88|81.82%|[226, 230, 231, 371, 372, 375, 376, 377, 378, 481, 482, 651, 681, 682, 701, 708]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/FragmentInstanceExecState.java|5|6|83.33%|[390]|
|:large_blue_circle:|com/starrocks/qe/scheduler/dag/ExecutionFragment.java|142|166|85.54%|[103, 107, 111, 127, 150, 157, 262, 263, 265, 266, 270, 271, 272, 276, 277, 278, 279, 293, 304, 305, 311, 341, 349, 353]|
|:large_blue_circle:|com/starrocks/qe/DefaultCoordinator.java|54|61|88.52%|[362, 643, 937, 938, 1055, 1141, 1388]|
|:large_blue_circle:|com/starrocks/qe/scheduler/TFragmentInstanceFactory.java|83|93|89.25%|[72, 121, 147, 148, 149, 150, 151, 173, 174, 175]|
|:large_blue_circle:|com/starrocks/qe/scheduler/Deployer.java|78|82|95.12%|[171, 172, 173, 203]|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|9|9|100.00%|[]|

**Review by liuyehcf — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by dirtysalt** [comment]
@mergify backport branch-3.1

**Pull request merged**